File: app/src/main/java/cn/devifish/readme/view/base/BaseRecyclerAdapter.java
package cn.devifish.readme.view.base;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.List;

/**
 * Created by zhang on 2017/4/15.
 *
 * @author zhang
 */

public abstract class BaseRecyclerAdapter<M, VH extends BaseViewHolder, L extends BaseViewHolder.OnItemClickListener> extends RecyclerView.Adapter<VH> {

    @NonNull private List<M> mList;
    @Nullable private L mListener;

    public BaseRecyclerAdapter(@NonNull List<M> list) {
        mList = list;
    }

    @Override
    public VH onCreateViewHolder(ViewGroup parent, int viewType) {return onCreateView(parent, viewType);}

    @Override
    public void onBindViewHolder(VH holder, int position) {onBindView(holder, position);}

    @Override
    public int getItemCount() {return mList.size();}

    protected abstract VH onCreateView(ViewGroup group, int viewType);
    protected abstract void onBindView(VH holder, int position);

    public void setOnItemClickListener(@Nullable L listener) {mListener = listener;}

    public L getItemClickListener() {return mListener;}

    public M getItem(int position) {return this.mList.get(position);}

    public void removeItem(int position) {
        mList.remove(position);
        notifyItemRemoved(position);
    }

    public void removeItem(M item) {
        int position = mList.indexOf(item);
        if (position != -1) removeItem(position);
    }

    public void addItem(M item, int position) {
        mList.add(position, item);
        notifyItemInserted(position);
    }

    public void addItem(M item) {addItem(item, getItemCount());}

}

File: app/src/main/java/cn/devifish/readme/view/base/BaseRecyclerAdapter.kt
package cn.devifish.readme.view.base

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup

/**
 * Created by zhang on 2017/4/15.

 * @author zhang
 */

abstract class BaseRecyclerAdapter<M, VH : BaseViewHolder<M>>() : RecyclerView.Adapter<VH>() {

    var data: MutableList<M>? = null
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    constructor(list : MutableList<M>) : this() {
        this.data = list
    }

    var listener: BaseViewHolder.OnItemClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH = onCreateView(parent, viewType)

    override fun onBindViewHolder(holder: VH, position: Int) = onBindView(holder, position)

    override fun getItemCount(): Int {
        if (data != null) {
            return data!!.size
        }
        return 0
    }

    protected abstract fun onCreateView(group: ViewGroup, viewType: Int): VH
    protected abstract fun onBindView(holder: VH, position: Int)

    fun setOnItemClickListener(listener: BaseViewHolder.OnItemClickListener?) {
        this.listener = listener
    }

    fun getItem(position: Int): M? {
        if (data != null) {
            return this.data!![position]
        }
        return null
    }

    fun removeItem(position: Int) {
        if (data != null) {
            data!!.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    fun removeItem(item: M) {
        if (data != null) {
            val position = data!!.indexOf(item)
            if (position != -1) removeItem(position)
        }
    }

    fun addItem(item: M, position: Int = itemCount) {
        if (data != null) {
            data!!.add(position, item)
            notifyItemInserted(position)
        }
    }

}
